Viet/script lookups

vvtran opened the pull request:
https://github.com/Plutonomicon/cardano-browser-tx/issues/122
- See below for comments on what this PR includes.

ngua reviewed (commented):
Here's one thing I think we'll need as well: at the moment, the constraints API is using `Types.Transaction.Redeemer`, right? But that's not the representation of a redeemer that users will interact with, so I think we need to define another redeemer type that's the same as the Plutus-style redeemer. Then we'll also need an effectful `convertRedeemer :: Redeemer -> QueryM Transaction.Redeemer` that:
- calculates the index (this corresponds to the index of the script in the inputs)
- calls the Haskell server for the ex unit calculation

vvtran commented:
> Here's one thing I think we'll need as well: at the moment, the constraints API is using `Types.Transaction.Redeemer`, right? But that's not the representation of a redeemer that users will interact with, so I think we need to define another redeemer type that's the same as the Plutus-style redeemer. Then we'll also need an effectful `convertRedeemer :: Redeemer -> QueryM Transaction.Redeemer` that:
> 
> * calculates the index (this corresponds to the index of the script in the inputs)
> * calls the Haskell server for the ex unit calculation

This is correct, I will add another wrapper over `PlutusData` and create a stub function which converts before calling the `attach_` FFIs.

ngua commented:
I should've mentioned that you can get the redeemer tag as well from the server call. So the client side will only have to fill in the `data` and `index` fields

vvtran commented:
> I should've mentioned that you can get the redeemer tag as well from the server call. So the client side will only have to fill in the `data` and `index` fields

I think this is possible with our current FFIs, we have: `foreign import getRedeemerTag :: Redeemer -> RedeemerTag`, if we want to "use less Haskell server"? We can then call `convertRedeemerTag :: RedeemerTag -> Maybe Tag.RedeemerTag`

vvtran commented:
Sorry, ignore my my deleted previous message - got confused.

ngua commented:
@vvtran Sorry, I was mistaken. We can get the index from the server call

vvtran commented:
- Created `ScriptLookups` API for creating lookups. Some functions aren't pure in comparison to their Plutus counterparts. We could potentially write a wrapper over `ScriptLookups` for the `Contract` monad so that its `Monoid` instance includes the effectful context and potential failure - so the user doesn't need to worry about dealing with effects and failure when combining with `<>`.
- Refactored various modules.
- Built helpers for getting hashes for pubkey, paymentpubkey, stakekey etc.
- Built helpers for getting hashes for `Validator`, `MintingPolicy` etc. (they're all wrappers around `PlutusScript` anyway).
- Built a function to build an unbalanced transaction given a set of lookups and constraints `mkUnbalancedTx` which comes from adding a `ReaderT` and `StateT` ontop of `QueryM` then running the stack back to `QueryM`.
- Built helpers for converting between `POSIXTimeRange` and `Slot` - again, some of the functions can fail, unlike their plutus counterparts.
- Provided 3 options for `OgmiosTxOut` which we already have defined but I wanted to follow the plutus API closely, mainly because chain index is often used in off-chain code, in particular, their lenses. Replicating this may be helpful for Haskell users. I suspect  Option 1 is tricky and therefore stick with Option 3 unless we convert for the `Contract` Monad. Anyhow, switching back to Option 3 should be easy, but please read through my comments on this.
- It was convenient to use Lenses for `processConstraints` here like Plutus so I built a small lens library. We of course have row polymorphism too so the user has options at least.
- A lot of these functions are related to https://github.com/Plutonomicon/cardano-browser-tx/issues/109
Outstanding:
- Add  (typeclass?) constraints for typed validators, input and output constraints -> add to `processLookupsAndConstraints`
- Complete redeemer conversion once Haskell server is ready https://github.com/Plutonomicon/cardano-browser-tx/issues/156

vvtran commented:
I also left a question around `ConstraintsM` if anyone has any thoughts, much appreciated!

ngua commented on src/Types/ScriptLookups.purs:
```diff
@@ -0,0 +1,1053 @@
+--------------------------------------------------------------------------------
+-- Create ScriptLookups helpers for `Contract` monad.
+--------------------------------------------------------------------------------
+-- FIX ME: Need to work out what to do with TypedValidator and constraints.
```
Maybe include the URL to the issue here (I believe it's been opened)

ngua commented on src/Types/ScriptLookups.purs:
```diff
@@ -0,0 +1,1053 @@
+  tupleToRecord (address /\ validator /\ datum /\ value) =
+    { address, validator, datum, value }
+
+-- -------------------------------- OPTION 2 --------------------------------------
```
I think we can get rid of all of this and just conclusively go with option 3

ngua commented on src/Types/ScriptLookups.purs:
```diff
@@ -0,0 +1,1053 @@
+    , value
+    }
+
+-- | Lenses to replicate the Plutus API
```
I don't think we need any of these if we're going with Option 3

ngua reviewed (commented):
This was a ton of work, thank you for taking care of this!

There are some more specific comments below, but more generally:
- I noticed quite a few TODOs, FIXMEs, etc... If there is something actionable _and_ crucial to be done, could you open an issue? In some other cases, we just need to choose one of several alternatives. In other cases, an explanatory comment  would suffice I think
- If there is a choice between a complex/non-idiomatic solution that preserves Plutus' API and a simpler one, we should opt for the latter. We already need to diverge significantly from the Plutus API due to certain limitations in PS (e.g. no type families)

vvtran commented on src/Types/Transaction.purs:
```diff
@@ -61,6 +69,31 @@ instance monoidTransaction :: Monoid Transaction where
     }
 
+--------------------------------------------------------------------------------
+-- `Transaction` Lenses
```
Happy to change snake case to camel case. Although I kept them as snake case for now simply to match the data types as they are. The reason is, when we finally convert our field names to camel case, it's a simple search and replace (on lenses too). However, if for example `validity_start_interval` gets renamed to `startInterval`, we may forget to rename the `_validityStartInterval` lens to `_startInterval`. I'm guessing a word change (regardless of case) is unlikely anyway so will rename them now.

vvtran commented on src/Types/ScriptLookups.purs:
```diff
@@ -0,0 +1,1053 @@
+  _unbalancedTx <<< _utxoIndex %= flip union txOutsMap
+
+data MkUnbalancedTxError
+  -- = TypeCheckFailed ConnectionError
```
Will remove, it's based on a typed validator error.

vvtran commented on src/Types/Interval.purs:
```diff
@@ -277,3 +296,184 @@ before h (Interval { ivFrom }) = lowerBound h < ivFrom
+  { validityStartInterval :: Maybe Slot, timeToLive :: Maybe Slot }
+
+-- FIX ME: I've interpreted `Nothing` as below, although this may change
+-- depending on our desired behaviour. I'm not entirely sure what why we have
```
I initially used that interpretation for positive infinity (for `ttl`). That gives two meanings to "max slot", either `Nothing` or `maxSlot`. Also, would that imply `validity_start_interval` can either be `Slot 0` or `Nothing` - they're essentially equivalent unless inclusiveness bounds are the difference...?

vvtran replied:
Made a change for this.

vvtran commented on src/Scripts.purs:
```diff
@@ -0,0 +1,90 @@
+plutusScriptHash = map (map wrap) <<< scriptHash <<< unwrap
+
+-- | Converts a `PlutusScript` to a `ScriptHash`.
+scriptHash :: PlutusScript -> Effect (Maybe ScriptHash)
```
It just occurred to me I can use `scriptHashFromBytes <<< unwrap`. I'm not sure why I went via Foreign `PlutusScripts`, my bad.

vvtran replied:
```
scriptHash :: PlutusScript -> Maybe ScriptHash
scriptHash = scriptHashFromBytes <<< unwrap
```
Everything in that module can only fail now (haven't pushed)

vvtran commented:
Tried to address your comments @ngua  with the exception of switching to Option 3 switch and Redeemer related, so leaving the conversations unresolved for now. Will keep the redeemer convo open until resolved in a separate PR. 

The rest of this PR will be changing to Option 3 + any further review changes.

vvtran commented:
I ended refactoring a few things and adding further helpers. I haven't pushed as it's not quite ready yet but should be for tomorrow.

vvtran commented on src/Types/Interval.purs:
```diff
@@ -277,3 +296,184 @@ before h (Interval { ivFrom }) = lowerBound h < ivFrom
+      , ivTo: UpperBound end endIncl
+      } = map (posixTimeToEnclosingSlot sc) ptr
+
+    closureWith
```
Sure.

vvtran commented on src/Types/Value.purs:
```diff
@@ -206,6 +243,21 @@ instance joinSemilatticeNonAdaAsset :: JoinSemilattice NonAdaAsset where
+  negation = wrap <<< map (map negate) <<< unwrap
+
+instance Split NonAdaAsset where
+  split (NonAdaAsset mp) = NonAdaAsset neg /\ NonAdaAsset nneg
```
Thanks!

vvtran commented on src/Types/Transaction.purs:
```diff
@@ -89,17 +122,17 @@ instance semigroupTxBody :: Semigroup TxBody where
           txB'.validity_start_interval
     , mint: txB.mint <> txB'.mint
-    , script_data_hash: txB.script_data_hash <<>> txB'.script_data_hash
+    , script_data_hash: txB.script_data_hash </> txB'.script_data_hash
```
Potentially a bit confusing but the old `<<>>` use to be `Last` also, I'm just renaming to something a bit nicer so that
`<\>` is `First`
`</> is `Last` (use to be  `<<>>`)
`<<>>` is the new `maybeArrayMerge`. 

Just wanted to make sure it was clear the logic is unchanged.

Secondly, are you wondering why we don't use `maybeArrayMerge`? I don't think it matters too much because we only want to append to `script_data_hash` via an effectful call. In particular, see 
`setScriptDataHash :: Array Redeemer -> Array Datum -> Transaction -> Effect Transaction`. Furthermore, I'm not sure if it makes much sense to mix two transaction `script_data_hash`es together, and this may hold for a few other fields, hence the use of `Last`.

If there is a particular reason we require merging, we can always switch.

adamczykm replied:
I like the naming changes a lot. I think I just don't grasp merging tx's together and how can some data from the first one discarded.
Also I don't like the effectful interface, but I guess we can address this in the long run.

vvtran replied:
I actually agree about merging tx. I think we were trying to replicate Plutus `tx`'s `Monoid` instance which may be deprecated now, and in the process, came across fields that were monoidally unintuitive in our `TxBody`. Interestingly `TxInfo` doesn't have a `Semigroup` instance. Perhaps it makes more sense to define an "empty" value, without semigroup or monoid and leave it at that.

vvtran commented on src/Types/PlutusData.purs:
```diff
@@ -86,24 +79,7 @@ instance DecodeAeson PlutusData where
-unitRedeemer :: RedeemerTag -> Redeemer
-unitRedeemer = undefined
-
+-- | Don't distinguish "BuiltinData" and "Data" like Plutus:
```
I've changed and removed the pipe.

vvtran commented:
@ngua  @adamczykm 
- Changed everything over to option 3.
- Fixed some logic because of `Maybe` issues in `processConstraints` for datums and datumhashing (see below).
- Refactored and move some files to `txOut` for conversions between inputs/output types. Noticed a bug related to the previous point. The `Maybe` context is a bit annoying to deal with in the various "datum hash" types. Sometimes you want `Nothing` to not fail because that represents a wallet address, but you do want failure if the `datumHash` function fails given you have a script address.
- Added a few more conversion functions in `txOut` so we can go between the various output types.
- Renamed `JsonWsp.Address` to `JsonWsp.OgmiosAddress` to prevent confusion.
- Refactored and added a few helpers to `Address`.
